src/number_theory/miller_rabin.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2022 Bolton Bailey, Sean Golinski. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey, Sean Golinski
+-/
+
+import data.fintype.basic
+import group_theory.order_of_element
+import tactic.zify
+import data.nat.totient
+import data.zmod.basic
+import number_theory.padics.padic_norm
+import field_theory.finite.basic
+import data.fintype.basic
+
+def two_power_part (n : ℕ) := 2 ^ (padic_val_nat 2 n)
+
+def odd_part (n : ℕ) := n / two_power_part n
+
+lemma mul_two_power_part_odd_part (n : ℕ) : (two_power_part n) * (odd_part n) = n :=
+begin
+  have : two_power_part n ∣ n,
+  { rw two_power_part,
+    exact pow_padic_val_nat_dvd, },
+  rw odd_part,
+  exact nat.mul_div_cancel' this,
+end
```

Comment:
I don't know if this helps, but you could re-write this in terms of `nat.factorization` (and in fact move it to `data/nat/factorization`) as follows:
````lean
noncomputable
def two_power_part (n : ℕ) := 2 ^ (n.factorization 2)

noncomputable
def odd_part (n : ℕ) := n / two_power_part n

lemma two_power_part_dvd (n : ℕ) : two_power_part n ∣ n :=
pow_factorization_dvd n 2

lemma mul_two_power_part_odd_part (n : ℕ) : (two_power_part n) * (odd_part n) = n :=
nat.mul_div_cancel' (two_power_part_dvd n)
````
In subsequent lemmas you could then either use `padic_val_nat_eq_factorization` to convert between `factorization` and `padic_val_nat`, or re-state these lemmas in terms of `n.factorization`.  Let me know if there's anything you'd need to be added to `data/nat/factorization` to make this work.

Reply:
I guess `padic_val_nat x y` is the same as `y.factorization x` for prime `x`. I guess I find `factorization` easier to understand, so I'm in favor of changing `padic_val_nat` to `factorization` here and anywhere else where `x` is prime. 

The only thing is, why is factorization noncomputable? I don't want to have to mark all my definitions as noncomputable just to switch the name.

Reply:
That's just because `factorization` is defined as a finsupp, and `data/finsupp/basic` is marked `noncomputable theory`.

Reply:
As @rwbarton suggests in [this thread](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/.60ite.60.20with.20multiple.20decidable.20instances/near/273154786), would making this file a `noncomputable theory` be compatible with your plans for this work?

Reply:
No, it's not compatible with our plans. We are going to want to evaluate `strong_probable_prime`, and it seems the algorithm provided by `padic_val_nat` allows us to do this.

Reply:
Fwiw, this seems somewhat related to [my unpolished code to factorize ints into powers of 4 / odd primes](https://github.com/Ruben-VandeVelde/flt/blob/030b00dfefa4eefbb4491c93c24f5d7450a36450/src/odd_prime_or_four.lean#L154); not sure if there's anything helpful there

Reply:
Now that #12301 is merged and `factorization` is computable, it should be possible to switch over from `padic_val_nat` to `factorization` now.